### PR TITLE
VulkanDeviceAlloc: Memorytype per slab

### PR DESCRIPTION
**Pull request by hrydgard**
See #11743 for the details. This fixes a wrong assumption we had about driver behavior.

### CODE REVIEW

**Comment by hrydgard** [comment]
@unknownbrackets  I'm gonna go ahead and merge this, since I think it's pretty safe. Let me know what you think when you have a sec.

File: Common/Vulkan/VulkanImage.cpp [hunk 1]
```diff
@@ -76,6 +77,8 @@ bool VulkanTexture::CreateDirect(VkCommandBuffer cmd, int w, int h, int numMips,
 	if (allocator_) {
 		offset_ = allocator_->Allocate(mem_reqs, &mem_, Tag());
 		if (offset_ == VulkanDeviceAllocator::ALLOCATE_FAILED) {
+			vkDestroyImage(vulkan_->GetDevice(), image_, nullptr);
```
Comment (unknownbrackets) [inline Common/Vulkan/VulkanImage.cpp]: Destroy is called in the destructor which already does this.  We should at least clear `image_`, but I'm personally not a big fan of repeating WET cleanup code in every error case, so I try to design things so they auto-cleanup and the code is easier to read.

-[Unknown]

File: Common/Vulkan/VulkanImage.cpp [hunk 2]
```diff
@@ -89,6 +92,7 @@ bool VulkanTexture::CreateDirect(VkCommandBuffer cmd, int w, int h, int numMips,
 
 		res = vkAllocateMemory(vulkan_->GetDevice(), &mem_alloc, NULL, &mem_);
 		if (res != VK_SUCCESS) {
+			ELOG("vkAllocateMemory failed: %s", VulkanResultToString(res));
 			_assert_msg_(G3D, res != VK_ERROR_TOO_MANY_OBJECTS, "Too many Vulkan memory objects!");
 			_assert_(res == VK_ERROR_OUT_OF_HOST_MEMORY || res == VK_ERROR_OUT_OF_DEVICE_MEMORY || res == VK_ERROR_TOO_MANY_OBJECTS);
 			vkDestroyImage(vulkan_->GetDevice(), image_, nullptr);
```
Comment (unknownbrackets) [inline Common/Vulkan/VulkanImage.cpp]: FWIW, this one is also not necessary but at least clears `image_` so is fine.

**Review by unknownbrackets — COMMENTED** [review]
Makes sense, hopefully it doesn't make a big difference on the size increase, depends how often this happens...

-[Unknown]

File: Common/Vulkan/VulkanImage.cpp (hunk 1 above)
Comment (hrydgard) [reply Common/Vulkan/VulkanImage.cpp]: WET?  Yeah I see your point, I'll clean it up.

Comment (unknownbrackets) [reply Common/Vulkan/VulkanImage.cpp]: We Enjoy Typing.  The opposite of DRY.

-[Unknown]

Comment (hrydgard) [reply Common/Vulkan/VulkanImage.cpp]: Ah :)

**Pull request merged**